app/routes/task_routes.py

```diff
@@ -1 +1,141 @@
+    if title_param:
+        query = query.where(Task.title.ilike(f"%{title_param}%"))
+    
+    description_param = request.args.get("description")
+    if description_param:
+        query = query.where(Task.description.ilike(f"%{description_param}%"))
+
+    completed_at_param =request.args.get("completed_at")
+    if completed_at_param:
+        query = query.where(Task.completed_at.ilike(f"%{completed_at_param}%"))
+
+    is_complete_param = request.args.get("is_complete")
+    if is_complete_param:
+        # is_complete_bool = is_complete_param.lower() == "true"
+        # query = query.where(Task.is_complete == is_complete_bool)
+
+        if is_complete_param.lower() == "true":
+            query = query.where(Task.task_completed == True)
+        elif is_complete_param.lower() == "false":
+            query = query.where(Task.task_completed == False)
+
+    query = query.order_by(Task.id)
+    tasks = db.session.scalars(query)
+
+    tasks_response = []
+    for task in tasks:
+        tasks_response.append(task.to_dict())
+    return tasks_response, 200
+
+
+@bp.get("/<task_id>")
+def get_one_task(task_id):
+    task = validate_model(Task, task_id)
+
+    return {"task": task.to_dict()}, 200
+
+@bp.patch("/<task_id>/mark_complete")
+def mark_task_as_complete(task_id):
+    task = validate_model(Task, task_id)
+
+    task.is_complete = True
+    task.completed_at = datetime.now(timezone.utc)
+    db.session.commit()
+
+    slack_token = os.environ.get("SLACKBOT_API_TOKEN")
+    slack_channel = os.getenv("SLACK_CHANNEL", "#sno-task-tracker-bot2")
+    slack_text = f"Someone just completed the task: *{task.title}* ! 🎉"
+
+    if slack_token:
+        url =  "https://slack.com/api/chat.postMessage"
+        headers = {
+            "Authorization": f"Bearer {slack_token}",
+            "Content-Type": "application/json"
+        }
+        slack_data = {
+            "channel": slack_channel,
+            "text": slack_text
+        }
+        slack_response = requests.post(url, headers=headers, json=slack_data)
+        
+        if not slack_response.ok:
+            print("Slack error:", slack_response.text)
+
+    response_body = {
+        "id": task.id,
+        "title": task.title,
+        "description": task.description,
+        "is_complete": task.is_complete,
+        "completed_at": task.completed_at.isoformat()
+    }
+
+    return make_response(response_body, 200)
+
+@bp.patch("/<task_id>/mark_incomplete")
+def mark_task_as_incomplete(task_id):
+    task = validate_model(Task, task_id)
+    
+    # task.is_complete = False
+    task.completed_at = None
+
+    db.session.commit()
+    return Response(status=204, mimetype="application/json")
+    # return {"task": task.to_dict()}, 200
+
+@bp.put("/<task_id>")
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
```

Comment:
This is another opportunity to D.R.Y. up our code! Notice how in this route and the `PUT` route for `goal_routes.py` we follow the pattern of:
```py
object.ATTRIBUTE = request_body["ATTRIBUTE"]
```
We use `hasattr` and [setattr](https://docs.python.org/3/library/functions.html#setattr) to make a helper function to update our `Task` and `Goal` model. It would look like this: 
```py
def update_model(obj, data):
    for attr, value in data.items():
        if hasattr(obj, attr):
            setattr(obj, attr, value)
    
    db.session.commit()

    return Response(status=204, mimetype="application/json")
```

This refactor not only makes our code D.R.Y but shows that we recognize logic that has higher level usability while handling cases of keys not being found!